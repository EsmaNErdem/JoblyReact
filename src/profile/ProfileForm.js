import React, { useState, useContext } from "react";
import UserContext from "../auth/UserContext";
import JoblyApi from "../api";
import Alert from "../utilities/Alert";

/**
 * ProfileForm Component
 * Displaya a form with nad rendered controlled component functionality
 * 
 * * A form component for user profile update. Handles user input for, first name, last name, and email while username constant.
 * Displays form errors if registration fails.
 * On submission:
 * - calls API to user info
 * - update currentUser state
 * - shows conformation message when update is successful
 *
 * Routes ==> ProfileForm ==> Alert
 * Routed as /profile
 */

const ProfileForm = () => {
    const { currentUser, setCurrentUser } = useContext(UserContext);

    const [formData, setFormData] = useState(
        {
            firstName: currentUser.firstName,
            lastName: currentUser.lastName,
            email: currentUser.email
        }
    )
    const [formErrors, setFormErrors] = useState([]);
    const [message, setMessage] = useState(false)

    console.debug(
        "ProfileForm",
        "currentUser=", currentUser,
        "formData=", formData,
        "formErrors=", formErrors,
    );

    /**
     * Handles form submission. Call updateUser API to update user data
     * Updates currentUser state with API call return updatedUser
     */
    const handleSubmit = async e => {
        e.preventDefault();
        let updatedUser;
        
        try {
            updatedUser =  await JoblyApi.updateUser(currentUser.username, formData)
        } catch (e) {
            setFormErrors(e)
        }

        setFormErrors([])
        setMessage(true)
        setCurrentUser(u => ({...u, updatedUser}))
    }

    /** Update form data field */
    const handleChange = e => {
        const { name, value } = e.target
        setFormData(data => ({
            ...data,
            [name] : value
        }))
    }

    return (
        <div>
            <h3>Profile Update</h3>
            <form onSubmit={handleSubmit}  className="SignUpForm">
                <label htmlFor="username">Username</label>
                <p>{currentUser.username}</p>
                <label htmlFor="firstName">First Name</label>
                <input
                    id="firstName"
                    name="firstName"
                    value={formData.firstName}
                    onChange={handleChange}
                    placeholder="First Name"
                />
                <label htmlFor="lastName">Last Name</label>
                <input
                    id="lastName"
                    name="lastName"
                    value={formData.lastName}
                    onChange={handleChange}
                    placeholder="Last Name"
                />
                <label htmlFor="email">Email</label>
                <input
                    id="email"
                    name="email"
                    type="email"
                    value={formData.email}
                    onChange={handleChange}
                    placeholder="email"
                />

                {formErrors.length
                    ? <Alert type="danger" messages={formErrors} />
                    : null
                }

                {message
                    ? <Alert type="success" messages={"Updated profile!"} />
                    : null
                }
                <button>Save Changes</button>
            </form>
        </div>
    )
}

export default ProfileForm;