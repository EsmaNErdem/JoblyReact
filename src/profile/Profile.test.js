import { render, fireEvent, waitFor } from "@testing-library/react";
import ProfileForm from "./ProfileForm";
import { UserProvider } from "../testUtils";

jest.mock('../api');

test("renders without crashing", async () => {
    render(
        <UserProvider>
            <ProfileForm />
        </UserProvider>
    ); 
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(
        <UserProvider>
            <ProfileForm />
        </UserProvider>
        );
    expect(asFragment()).toMatchSnapshot();
});

test("LoginForm renders correctly", () => {
    const { getByLabelText, getByText } = render(
        <UserProvider>
            <ProfileForm />
        </UserProvider>
    );
    
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email");
    const saveButton = getByText("Save Changes");
  
    expect(firstNameInput).toBeInTheDocument();
    expect(lastNameInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(saveButton).toBeInTheDocument();
  });
  
  test("Submitting the form calls login function", async () => {
    const { getByLabelText, getByText } = render(
          <UserProvider >
              <ProfileForm />
          </UserProvider>
      );
    
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email");
    const saveButton = getByText("Save Changes");
  
    fireEvent.change(firstNameInput, { target: { value: "user" } });
    fireEvent.change(lastNameInput, { target: { value: "name" } });
    fireEvent.change(emailInput, { target: { value: "user@email.com" } });
    fireEvent.click(saveButton);

    await waitFor(() => {
        expect(getByText("Updated profile!")).toBeInTheDocument();
    });
  });
   
