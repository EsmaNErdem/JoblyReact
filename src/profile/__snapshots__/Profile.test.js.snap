// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submitting the form calls login function 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Profile Update
    </h3>
    <form
      class="SignUpForm"
    >
      <label
        for="username"
      >
        Username
      </label>
      <p>
        testuser
      </p>
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="First Name"
        value="user"
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        value="name"
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        placeholder="email"
        type="email"
        value="user@email.com"
      />
      <button>
        Save Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Profile Update
    </h3>
    <form
      class="SignUpForm"
    >
      <label
        for="username"
      >
        Username
      </label>
      <p>
        testuser
      </p>
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="First Name"
        value=""
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        value=""
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        placeholder="email"
        type="email"
        value="test@test.net"
      />
      <button>
        Save Changes
      </button>
    </form>
  </div>
</DocumentFragment>
`;
